contracts/SunDAI.sol

```diff
@@ -0,0 +1,263 @@
+import "../node_modules/openzeppelin-solidity/contracts/access/roles/MinterRole.sol";
+
+
+contract SunDAI is IERC20, MinterRole {
+  using SafeMath for uint256;
+  
+  bytes32 public name;
+  bytes32 public symbol;
+  uint8 public decimals;
+
+  address public bridgeAddr;
+  IERC20 public dai;
+  mapping(address => uint256) public daiBalance;
+
+  constructor(address _daiAddr, address _bridgeAddr, bytes32 _name, bytes32 _symbol) public {
+    dai = IERC20(_daiAddr);
+    bridgeAddr = _bridgeAddr;
+    decimals = 18;
+    name = _name;
+    symbol = _symbol;
+  }
+
+  // 1. DAI is pulled into the contract, and record created
+  // 2. sunDAI are minted in amount of DAI deposited
+  // 3. sunDAI transfered to the Bridge
+  function _pullAndMint(address from, address to, uint256 value) internal {
+    require(dai.transferFrom(from, address(this), value));
+    daiBalance[from] = daiBalance[from].add(value);
+    _mint(to, value);
+  }
+
+  function _transferFrom(address from, address to, uint256 value) internal {
+    _transfer(from, to, value);
+
+    // only check allowance if other caller than bridge
+    if (to != bridgeAddr) {
+      approve(from, _allowed[from][msg.sender].sub(value));
+    }
+  }
+
+  // +------+         +--------+       +--------+
+  // |  DAI |         | sunDAI |       | Bridge |
+  // +---+--+         +----+---+       +----+---+
+  //     |                 |                |
+  //     |                 |                |
+  //     |                 | transferFrom   |
+  //     |  transferFrom   | <------------+ |
+  //     | <------------+  |                |
+  //     |                 | mint           |
+  //     | +------------>  | +--+           |
+  //     |                 |    |           |
+  //     |                 | <--+           |
+  //     |                 |                |
+  //     |                 | transfer       |
+  //     |                 | +----------->  |
+  //     |                 |                |
+  function transferFrom(address from, address to, uint256 value) public returns (bool) {
+    uint256 amountToMint = 0;
+    if (_balances[from] < value) {
+      amountToMint = value.sub(_balances[from]);
+      if (amountToMint > 0) {
+        require(msg.sender == bridgeAddr, "pullAndMint only for bridge");
+        require(to == bridgeAddr, "pullAndMint only to bridge");
+      }
+      _pullAndMint(from, msg.sender, amountToMint);
+    }
+    if (amountToMint < value) {
+      if (msg.sender == bridgeAddr || to == bridgeAddr) {
+        require(msg.sender == bridgeAddr, "transferFrom only by bridge to bridge");
+        require(to == bridgeAddr, "transferFrom only by bridge to bridge 2");
+      }
+      _transferFrom(from, to, value.sub(amountToMint));
+    }
+    return true;
+  }
+
+  function burn(address owner) public {
+    require(daiBalance[owner] > 0);
+    uint256 amount = _balances[owner];
+    if (daiBalance[owner] < amount) {
+      amount = daiBalance[owner];
```

Comment:
We only burn the amount of DAI the owner has?
Shouldn't we burn every sunDAI and transfer all remaining DAI to the owner?

Reply:
If there are some SunDAI left you can sell them, so some-one else with a DAI allowance can burn them. Otherwise you end up with a contract that has > 0 Dai locked, but 0 total supply and no way to unlock the DAI.

I guess the problem is the name, I will consider renaming the function to `unlock`.

Reply:
Maybe do it like this:
if owner == 0: try to burn for msg.sender
if owner !=0: check if msg.sender is owner/admin of contract > burn the given owner addr if so